packages/sdk-client/src/builder/ApiRootBuilder.ts

```diff
@@ -0,0 +1,131 @@
+import fetch from 'node-fetch'
+import { ApiRoot, createApiBuilderFromCtpClient } from '@commercetools/platform-sdk'
+import {
+  Nullable,
+  Middleware,
+  Credentials,
+  AuthMiddlewareOptions,
+  HttpMiddlewareOptions,
+  UserAgentMiddlewareOptions,
+  QueueMiddlewareOptions,
+  CorrelationIdMiddlewareOptions
+} from '../types/sdk.d'
+import { default as createClient } from '../sdk-client/client'
+import { default as createHttpMiddleware } from '../sdk-middleware-http/http'
+import { default as createAuthForClientCredentialsFlow } from '../sdk-middleware-auth/client-credentials-flow'
+
+import { default as createUserAgentMiddleware } from '../sdk-middleware-user-agent/user-agent'
+import { default as createQueueMiddleware } from '../sdk-middleware-queue/queue'
+import { default as createLoggerMiddleware } from '../sdk-middleware-logger/logger'
+import { default as createCorrelationIdMiddleware } from '../sdk-middleware-correlation-id/correlation-id'
+
+export default class ApiRootBuilder {
+  private projectKey: string | undefined
+  private authMiddleware: Nullable<Middleware>;
+  private httpMiddleware: Nullable<Middleware>;
+  private userAgentMiddleware: Nullable<Middleware>;
+  private correlationIdMiddleware: Nullable<Middleware>;
+  private loggerMiddleware: Nullable<Middleware>;
+  private queueMiddleware: Nullable<Middleware>;
+  private middlewares: Array<Middleware> = []
+
+  withProjectKey(key: string): ApiRootBuilder {
+    this.projectKey = key;
+    return this;
+  }
+
+  defaultClient(baseUri: string, credentials: Credentials, oauthUri?: string, projectKey?: string): ApiRootBuilder {
+    return this.withAuthMiddleware({
+      host: oauthUri,
+      projectKey: projectKey || this.projectKey,
+      credentials
+    })
+      .withHttpMiddleware({
+        host: baseUri,
+        fetch: fetch,
+      })
+      .withLoggerMiddleware()
+  }
+
+  withAuthMiddleware(options: AuthMiddlewareOptions): ApiRootBuilder {
```

Comment:
i would introduce two methods here

`withAuthMiddleware(authMiddleware: Middleware)`

and `withClientCredentialsFlow(options: AuthMiddlewareOptions)`

cause this method atm only supports client credentials flow. For other auth types a specific authMiddleware has to be added.

Reply:
I have included other authentication middleware methods `withClientCredentialsFlow`, `withPasswordFlow`, `withAnonymousSessionFlow` `withRefreshTokenFlow` etc

Now, you can use any of the 5 client authentication method.